### PR TITLE
Add optional entropy-based early exit diagnostics

### PR DESCRIPTION
**Pull request by Moviw**
## Summary

  This PR adds an optional entropy-based early-exit mechanism to CTM inference.

  The feature uses the existing CTM certainty signal, 1 - normalized_entropy(logits), to record the first internal tick where each sample crosses a user-provided
  threshold. It preserves the default CTM forward behavior: when early_exit_threshold=None and return_exit_steps=False, return values and output shapes are unchanged.

  When enabled, finished samples have their prediction/certainty outputs frozen across later ticks. The recurrent loop only breaks once every sample in the batch has
  crossed the threshold, so the reported reduction is a per-sample mean exit tick diagnostic, not a claim of per-sample FLOP elimination.

  ## Motivation

  CTMs naturally emit predictions and certainty estimates at each internal thought tick. This makes them a useful setting for adaptive inference diagnostics without introducing a learned halting head or changing training.

  ## Results

  On a lightweight MNIST CTM:

| Threshold | Accuracy | Mean Exit Tick |
|---|---:|---:|
| baseline | 92.69% | 30.00 |
| 0.70 | 92.69% | 19.42 |
| 0.50 | 92.53% | 16.40 |

  ## Compatibility

  - Default forward behavior is unchanged.
  - Existing track=True behavior is unchanged unless return_exit_steps=True.
  - No new dependencies are added.

  ## Validation

  - python -m py_compile models/ctm.py models/ctm_sort.py tasks/image_classification/analysis/run_mnist_act_metrics.py
  - python -m tasks.image_classification.analysis.run_mnist_act_metrics

### CODE REVIEW

**Comment by lukedarlow** [comment]
I appreciate your contribution here, but I think that it is, unfortunately, not useful. I had implemented and tested this in early renditions of the CTM. The biggest benefit of something like this would be early exit for a subset of samples in a minibatch. However, whether considering training or inference, the compute graph afforded by such a fine-grained approach necessarily does not emit any neat compilations, meaning that you usually get a slowdown in performance. 

However counterintuitive this might seem (because early exiting like you've built it is certainly the human and intuitive way to do this), GPUs are not designed for this. A lot more effort, using sparse tensors and/or custom cuda kernels, would be needed to yield gains in this case. 

If you want to go down that route, I am happy to re-evaluate any performance gains you might present in the future. 

-------------
Caveat: if your intention is to simple "get the prediction" at some threshold for each sample, your technique here is not the canonical pythonic way of approaching this task. There are a few simple python lines to compute these exact quantities that do not require intervening in the forward pass.
